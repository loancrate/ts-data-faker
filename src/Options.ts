export interface Options {
  optionalRatio: number;
  arrayLengthMinimum: number;
  arrayLengthMaximum: number;
  pastDateYears: number;
  referenceDate: Date;
  warning(message: string): void;
}

export const defaultOptions: Options = {
  optionalRatio: 1,
  arrayLengthMinimum: 1,
  arrayLengthMaximum: 5,
  pastDateYears: 10,
  referenceDate: new Date(),
  // istanbul ignore next
  warning(message) {
    console.error("Warning:", message);
  },
};
