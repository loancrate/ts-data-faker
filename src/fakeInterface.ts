import { InterfaceDeclaration, Node } from "ts-morph";
import { fakeTypeElementMemberedNode } from "./fakeTypeElementMemberedNode";
import { ObjectContext } from "./ObjectContext";
import { Options } from "./Options";

export function fakeInterface(params: {
  node: InterfaceDeclaration;
  propertyPath?: string[];
  targetObject?: Record<string, any>;
  objectContext?: ObjectContext;
  options: Options;
}): Record<string, unknown> {
  const { node, propertyPath = [], targetObject = {}, objectContext = new ObjectContext(), ...rest } = params;
  for (const baseType of node.getBaseTypes()) {
    const declarations = baseType.getSymbolOrThrow().getDeclarations();
    for (const declaration of declarations) {
      /* istanbul ignore else */
      if (Node.isInterfaceDeclaration(declaration)) {
        fakeInterface({ node: declaration, propertyPath, targetObject, objectContext, ...rest });
      }
    }
  }
  return fakeTypeElementMemberedNode({ node, propertyPath, targetObject, objectContext, ...rest });
}
