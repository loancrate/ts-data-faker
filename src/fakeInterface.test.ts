import { faker } from "@faker-js/faker";
import { readFileSync } from "fs";
import { dirname, join } from "path";
import { Project } from "ts-morph";
import { fakeInterface } from "./fakeInterface";
import { defaultOptions, Options } from "./Options";

const testData = join(dirname(__dirname), "test-data");

describe("fakeInterface", () => {
  test("coverage", () => {
    const project = new Project({ useInMemoryFileSystem: true });
    const source = project.createSourceFile("index.ts", readFileSync(join(testData, "test.ts"), { encoding: "utf8" }));
    const node = source.getInterfaceOrThrow("Profile");
    faker.seed(42);
    const warning = jest.fn(/* (message) => console.log(message) */);
    const options: Options = {
      ...defaultOptions,
      arrayLengthMinimum: 2,
      optionalRatio: 0.5,
      referenceDate: new Date("2022-08-18T00:00:00Z"),
      warning,
    };
    const obj = fakeInterface({ node, options });
    expect(obj).toMatchSnapshot();
    expect(warning).toHaveBeenCalledWith("Unhandled type unknown for alwaysUnknown");
    expect(warning).toHaveBeenCalledWith("Unhandled FunctionType declaration for fn");
  });
});
