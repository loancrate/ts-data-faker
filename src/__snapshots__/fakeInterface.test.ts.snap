// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`fakeInterface coverage 1`] = `
{
  "__typename": "Profile",
  "accounts": [
    {
      "accountNumber": "56795372",
      "balance": 4389714.21,
      "balanceDate": 2013-06-02T08:46:44.341Z,
      "bureau": "Bechtelar - Lemke",
      "name": "because oh fooey",
      "type": "SAVINGS",
    },
    {
      "accountNumber": "75642370",
      "balance": 1160726.4,
      "balanceDate": 2013-02-03T21:49:53.142Z,
      "bureau": "Okuneva, Goyette and Wilkinson",
      "name": "dime",
      "type": "SAVINGS",
    },
    {
      "accountNumber": "03658610",
      "balance": 6424192.78,
      "balanceDate": 2012-11-24T18:23:26.865Z,
      "bureau": "Morissette LLC",
      "name": "controvert",
      "type": "CHECKING",
    },
    {
      "accountNumber": "08216687",
      "balance": 8034809.31,
      "balanceDate": 2015-06-15T10:13:42.524Z,
      "bureau": "Bernier, Hamill and Sporer",
      "name": "waterlogged adventurously",
      "type": "CHECKING",
    },
  ],
  "age": 31,
  "alwaysNull": null,
  "amount": 8074401.55,
  "apr": 4.27,
  "birthDate": 1947-03-17T05:05:57.599Z,
  "city": "Donnellytown",
  "count": 9,
  "countryName": "Botswana",
  "countyFipsCode": "37684",
  "countyName": "Union County",
  "email": "Virgil_Haag@gmail.com",
  "firstName": "Virgil",
  "fullName": "Virgil Haag",
  "id": "5fb9220d-9b0f-4d32-a248-6492457c3890",
  "identifier": "6YRXWL",
  "ipAddress": "74.45.38.182",
  "language": "Latin",
  "lastName": "Haag",
  "middleName": "Austin",
  "n": 818,
  "optional1": true,
  "pct": 0.2279,
  "phoneExtension": "72",
  "phoneNumber": "857.554.0106 x3592",
  "productOrAccount": {
    "description": "Tantillus tego amet veniam despecto trepide.",
    "details": "virga nam ratione",
    "price": 108376.51,
    "priceInBulk": 9053819.77,
    "productCode": "RTZEDR",
    "productName": "Gorgeous Steel Soap",
    "states": [
      "SOLID",
      "GAS",
    ],
    "status": "Glittering",
    "type": "UnlinedGradient",
  },
  "products": [
    {
      "description": "Convoco volup vivo caste derideo coerceo cinis adfectus.",
      "details": "voro carbo spectaculum",
      "price": 2786464.64,
      "priceInBulk": 9082658.86,
      "productCode": "0IF74C",
      "productName": "Tasty Concrete Keyboard",
      "states": [
        "SOLID",
        "LIQUID",
      ],
      "status": "Pink",
      "type": "MeagerAtrium",
    },
    {
      "description": "Audentia adhaero officiis spiculum acceptus.",
      "details": "amplitudo vesper utpote",
      "price": 1743664.29,
      "priceInBulk": 6909377.38,
      "productCode": "R8QDMM",
      "productName": "Licensed Bamboo Soap",
      "states": [
        "GAS",
        "SOLID",
        "LIQUID",
      ],
      "status": "Messy",
      "type": "EmptyParsnip",
    },
    {
      "description": "Confido conscendo sursum velit valens termes sapiente ago.",
      "details": "solio caelestis summa",
      "price": 91970.51,
      "priceInBulk": 1014715.42,
      "productCode": "9NTJJ8",
      "productName": "Electronic Steel Table",
      "states": [
        "SOLID",
        "LIQUID",
        "GAS",
      ],
      "status": "Damaged",
      "type": "UnrealisticNoon",
    },
    {
      "description": "Careo voluptate crudelis vapulus quos.",
      "details": "vilicus viriliter vergo",
      "price": 4925176.94,
      "priceInBulk": 1952429.87,
      "productCode": "8BQNUN",
      "productName": "Luxurious Bamboo Computer",
      "states": [
        "SOLID",
        "LIQUID",
      ],
      "status": "Stained",
      "type": "MeagerMinor",
    },
    {
      "description": "Comes assentator iusto video subito molestiae alter possimus vulariter.",
      "details": "usque vereor dolorem",
      "price": 7407686.18,
      "priceInBulk": 6970157.41,
      "productCode": "D0XFYY",
      "productName": "Small Concrete Fish",
      "states": [
        "LIQUID",
        "SOLID",
        "GAS",
      ],
      "status": "Cooperative",
      "type": "MintySwanling",
    },
  ],
  "score": 360,
  "stateCode": "SD",
  "street1": "13328 Goyette Mill",
  "street2": "Apt. 809",
  "suffixName": "MD",
  "taxpayerIdentification": {
    "value": "198688185",
  },
  "text": "Compono spectaculum tantum textilis testimonium. Depulso adsuesco inflammatio. Valens consequatur angelus aranea temptatio quae. Ago substantia aequus.",
  "title": "Product Operations Designer",
  "username": "Virgil.Haag12",
  "value": "traho suffoco aggredior",
  "zipCode": "77031-8630",
}
`;
